### PR TITLE
Ports React [READY]

### LINKED ISSUE #78221: tgui - inferno linkEvents aren't compatible with Button components despite what the tgui-bench would make you think
## Round ID:
N/A, local server with a recent master

## Testmerges:
N/A

## Reproduction:
Have a tgui `Button` with a `linkEvent()` `onClick` event handler much like the one in the test bench
Results: clicking on the button results in an error
Expected: the handler being called just the same as if the event was bound without linkEvents.

Second Button works #1, so does the DOM input, but the first Button has the following error
image
image
https://github.com/tgstation/tgstation/blob/ff5695b7b8b260e29eb0e764e7cdb51b4212b8be/tgui/packages/tgui/components/Button.js#L53

The bench test in question. Sure, it renders fine and fast, but clicking on it would totally error out!
https://github.com/tgstation/tgstation/blob/ff5695b7b8b260e29eb0e764e7cdb51b4212b8be/tgui/packages/tgui-bench/tests/Button.test.tsx#L21

I imagine the fact that linkEvents work fine if bound directly on the `onClick` of a DOM element indicates that they're not too far off from working and that the solution is to handle `onClick` potentially being a `linkEvent` object rather than a function in `Button.js`

linkEvents are recommended in the [Inferno.js quick start guide](https://www.infernojs.org/docs/guides/event-handling).
>linkEvent() offers better performance than binding an event in a class constructor and using arrow functions, so use it where possible.

### PR DESCRIPTION
**Pull request by jlsnow301**
## About The Pull Request
Moves us from Inferno 7.4.11 to React 18.2. This is a big project! You might find errors here. Sometimes I get carried away fixing  external issues, I'll try to atomize them into separate PRs.

- [x] Implements React
- [x] CSS kebab to camel case
- [x] Some typescript refactors 
- [x] Atomize what I can
- [x] Fix dropdowns (opens in the wrong location)
- [x] Fix wonky button placement in a few UIs
- [x] Cleanup
## Why It's Good For The Game
- It'd be nice to just have real hooks. Our implementation is probably the best possible for Inferno, don't get me wrong.
- React is just easier to maintain and easier to read. It's the web dev standard.

Closes #78221 
## Changelog
:cl:
refactor: TGUI V5: The UI has had its entire engine replaced with React v18.2. This might cause obvious or laughably broken UIs in places you wouldn't expect. Please report any issues you find to the repo!
/:cl:

### CODE REVIEW

**Comment by SyncIt21** [comment]
With all inferno components now replaced with react, Can we close?
- #78221
- #78220

**Comment by Mothblocks** [comment]
I suspect only the first of those can be closed

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: tgui/packages/tgui/interfaces/OreRedemptionMachine.jsx [hunk 1]
```diff
@@ -181,8 +181,7 @@ const MaterialRow = (props) => {
             height="18px"
             width="18px"
             style={{
-              '-ms-interpolation-mode': 'nearest-neighbor',
-              'vertical-align': 'middle',
+              verticalAlign: 'middle',
```
[the remark this answers is not in the document]
Comment (jlsnow301) [reply tgui/packages/tgui/interfaces/OreRedemptionMachine.jsx:184]: i've added a temporary bypass - all Boxes used as="img" will add the style.

File: tgui/packages/tgui/components/Popper.tsx [hunk 2]
```diff
@@ -43,7 +45,9 @@ export class Popper extends Component<PopperProps> {
       // This code is copied from `findDOMNode` in inferno-extras.
       // Because this component is written in TypeScript, we will know
       // immediately if this internal variable is removed.
-      const domNode = findDOMfromVNode(this.$LI, true);
+      //
+      // eslint-disable-next-line react/no-find-dom-node
+      const domNode = findDOMNode(this) as Element;
```
Comment (Mothblocks) [inline tgui/packages/tgui/components/Popper.tsx:50]: Either make this use refs or update the comment to remove the mention of Inferno

Comment (jlsnow301) [reply tgui/packages/tgui/components/Popper.tsx:50]: refs are producing the same error unfortunately

File: tgui/packages/tgui/interfaces/DepartmentOrders.tsx [hunk 3]
```diff
@@ -116,7 +116,7 @@ const DepartmentCatalog = (props) => {
         <Tabs textAlign="center" fluid>
           {supplies.map((cat) => (
             <Tabs.Tab
-              key={cat}
+              key={cat.name}
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/DepartmentOrders.tsx:119]: Using an object as a key :|

File: tgui/packages/tgui/interfaces/FishingRod.tsx [hunk 4]
```diff
@@ -67,10 +61,10 @@ const FishingRodSlot = (props: FishingSlotProps) => {
 };
 
 export const FishingRod = (props) => {
-  const { act, data } = useBackend<FishingRodData>();
+  const { data } = useBackend<FishingRodData>();
 
   return (
-    <Window>
+    <Window height={300} width={300}>
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/FishingRod.tsx:67]: Window must have dimensions set, this caused #63498

File: tgui/packages/tgui/components/Box.tsx [hunk 5]
```diff
@@ -202,44 +194,42 @@ const styleMapperByPropName = {
...
-};
+} as const;
+
+// Boolean props
+const booleanStyleMap = {
+  bold: mapBooleanPropTo('fontWeight', 'bold'),
+  inline: mapBooleanPropTo('display', 'inline-block'),
+  italic: mapBooleanPropTo('fontStyle', 'italic'),
+  nowrap: mapBooleanPropTo('whiteSpace', 'nowrap'),
+  preserveWhitespace: mapBooleanPropTo('whiteSpace', 'preWrap'),
```
Comment (jlsnow301) [inline tgui/packages/tgui/components/Box.tsx:205]: Likely cause of #80100

**Pull request merged**